### PR TITLE
Stateful metrics

### PR DESCRIPTION
**Pull request by abhmul**
This pull request mainly focuses of a new API for metrics that allows them to be stateful. This means, metrics are more flexible and can persist state between batches. This PR will only support metric state preserved through batches. At the end of each epoch, the metric is reset back to its initial state. Below is a more thorough description and explanation of stateful metrics:

# Stateful Metrics 

An abstract metric class `Metric` is defined in `pyjet/abstract_metrics.py`. Some notes on it:

- Passing a function of the form `metric(y_pred, y_true)` to an abstract metric will use that function to calculate the score on a batch.
- The accumulate method is called at the end of each batch to calculate the aggregate score over the entire epoch thus far (see `AverageMetric` for an example what an accumulate method might look like).
- The reset method is called at the end of each epoch or validation run. It simply overwrites the attributes of the metric with its attributes at initialization.

Metrics are callable like any function and take as input:
```
batch_score = metric(y_pred, y_true)
```
where `y_true` are the labels for the batch and `y_pred` are the predictions

To implement your own custom metric, override the `score` method and the `accumulate` method. If you just want to average the scores over the epoch, consider using `AverageMetric` and just overriding the `score` function. `AverageMetric` already has its own `accumulate` method and will perform a weighted average the scores during training weighted by batch size.

Currently there are 3 pre-defined metrics that inherit `AverageMetric`. These are:

- `Accuracy`
- `AccuracyWithLogits`
- `TopKAccuracy`

These metric's function should be pretty self-explanatory. If you want to learn more about these functions, see their respective documentation. Note that `TopKAccuracy` has an additional initialization parameter:

```
TopKAccuracy(k=3)
```
differing from most metrics, which don't need any initialization parameters. 

## Registry

To make using predefined metrics a little easier, a registry was included in `pyjet/registry.py`. This file includes a dictionary mapping of metric names to their respective metric functions and includes the functions

```
register_metric(name, metric)
load_metric(name)
```

- `register_metric` is used to register a metric with its given name. Simply import this function into the file with metric and call it at the end of your file. Make sure the second argument is an *initialized* metric object.
- `load_metric` is used by `SLModel` to turn metric names into their respective metric functions.

This registry currently only stores metrics, but will be used for more in the future.

# Changes to Training

To accommodate these new changes to metrics for training and validating models, several methods had to be modified on `SLModel`. 

## New logging

One major addition was the new `TrainingLogs` class. This logger is a little similar to the old `MetricLogs` on the surface, but operates a little differently. Below are some notes on it

- Like the Keras logs, validation scores are now stored in the same logs as training scores, but are just prefixed with "val_".
- The logs maintains sub-logs for the current training epoch and training batch. The batch logs are just the scores on the current batch. The epoch logs are the scores accumulated over the entire epoch thus far. The accumulation is defined by each metric.
- The logs come with a few methods:

```
log_metric(metric, score)
```
This method simply updates the batch log with new score for the metric. It will also accumulate the score for the epoch logs

```
log_validation_metric(metric)
```
This method will accumulate the metric and store it prefixed with "val_". This is used during the validation portion of training.

In addition `on_epoch_begin` and `on_epoch_end` are use for some boilerplate and teardown. The logs bear some resemblance to `Callback`. The goal is to eventually turn logging and all non-optimization portions of training into their own callbacks and clean up the `fit` method.

## New Callbacks API

Because of the new logging classes, the API of callbacks was changed to look much more like the Keras API for callbacks. The changes to the API are pretty self-explanatory, so I won't go into detail, but you can see the diff for more details. 

One major addition was the `ProgressBar` callback. The old progress bar used during training, evaluation, and prediction when `verbose` was not 0 has been replaced with a callback.

Keep in mind that not all callbacks have been updated to the new API. I'll include these updates in another PR. Currently only `Plotter`, `ModelCheckpoint`, and `ProgressBar` work with the new API.

In addition, similar to Keras, a new container for a list of callbacks is being used called `CallbacksList`. This pretty much behaves as a composite callback that also performs some timing of callbacks. It'll issue a warning when the callbacks take too much time from training.

# Other Updates

- Since `MetricLogs` and `ProgBar` have been replaced and are no longer used, I deleted them.
- The MNIST example has been updated to work all the new changes. Plan is to eventually integrate it into the testing suite as a general training and prediction integration test. For now, just manually run it to test and update it when you make any breaking changes to the API.
- Multiple new tests have been added. Test coverage is still low, but its better than nothing.

### CODE REVIEW

**Comment by abhmul** [comment]
[Make stateful metrics](https://trello.com/c/KXL3bd6g/16-make-stateful-metrics)
